### PR TITLE
chore: use env https_proxy as oss client option proxy_host #12313

### LINKED ISSUE #12313: env http_proxy in wait/init container does not work for oss client
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

in my `workflow-controller-configmap` had configure `http_proxy` `https_proxy`
and it can access `gcs` `s3` artifact repository correctly, except for `oss` artifact repository

I created a test pod with setting the same proxy, I found that the oss-cli cannot not access the oss either
but I can access the oss bucket after applying the args `--proxy-host`
```
# use oss-cli to list buckets 
[root@test /]# aliyun oss ls
^C

# curl 
[root@test /]# curl -I http://${my-oss-bucket}.aliyuncs.com
HTTP/1.1 200 OK
Server: AliyunOSS
Date: Mon, 27 Nov 2023 08:13:57 GMT
Content-Length: 0
x-oss-request-id: xxx
x-oss-bucket-region: xxx
x-oss-server-time: 11
X-Cache: MISS from mail101
X-Cache-Lookup: HIT from mail101:3128
Via: 1.1 mail101 (squid)
Connection: keep-alive

# use oss-cli to list buckets with setting proxy args
[root@test /]# aliyun oss ls oss://${my-oss-bucket} --proxy-host ${my_proxy}
LastModifiedTime                   Size(B)  StorageClass   ETAG                                  ObjectName
2023-11-21 10:00:11 +0000 UTC          130      Standard   XXX      ${my-oss-bucket}/test.tgz
Object Number is: 1
```
Ref: [aliyun http proxy](https://help.aliyun.com/document_detail/121199.html)

The oss client in the argo code seems in the same case. The proxy can only be set through `oss.ClientOption`
However, we cannot configure the proxy for oss client through pod env or `argoexec` args currently
```
func (ossDriver *ArtifactDriver) newOSSClient() (*oss.Client, error) {
	var options []oss.ClientOption
        options = append(options, oss.Proxy("${my_proxy}")) <--- here
	if token := ossDriver.SecurityToken; token != "" {
		options = append(options, oss.SecurityToken(token))
	}
	client, err := oss.New(ossDriver.Endpoint, ossDriver.AccessKey, ossDriver.SecretKey, options...)
	if err != nil {
		return nil, fmt.Errorf("failed to create new OSS client: %w", err)
	}
	return client, err
}
```

### Version

v3.4.8 & v3.4.14

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
N/A
```

### Logs from the workflow controller

```text
N/A
```

### Logs from in your workflow's wait container

```text
time="2023-12-02T11:58:27 UTC" level=info msg="Save artifact" artifactName=main-logs duration=37.278144ms error="<nil>" key=2023/12/02/examples.use-cloud-acc-putb
j8br/examples.use-cloud-acc-putbj8br-3396519514/main.log
time="2023-12-02T11:58:27 UTC" level=info msg="not deleting local artifact" localArtPath=/tmp/argo/outputs/logs/main.log
time="2023-12-02T11:58:27 UTC" level=info msg="Successfully saved file: /tmp/argo/outputs/logs/main.log"
time="2023-12-02T11:58:27 UTC" level=info msg="Alloc=11282 TotalAlloc=17790 Sys=33661 NumGC=5 Goroutines=7"
time="2023-12-02T11:58:27 UTC" level=fatal msg="timed out waiting for the condition: Put \"http://${my-oss-bucket}.aliyuncs.com/aaa.tgz\": dia
l tcp 47.75.18.199:80: i/o timeout"
```

### PR DESCRIPTION
**Pull request by jingkkkkai**
Fixes #12313

### CODE REVIEW

**Review by terrytangyuan — COMMENTED** [review]
@shuangkun Could you help review this?

**Comment by shuangkun** [comment]
> @shuangkun Could you help review this?

Okay

File: workflow/artifacts/oss/oss.go [hunk 1]
```diff
@@ -84,6 +84,11 @@ func (p *ossCredentialsProvider) GetCredentials() oss.Credentials {
 
 func (ossDriver *ArtifactDriver) newOSSClient() (*oss.Client, error) {
 	var options []oss.ClientOption
+
+	if proxy, ok := os.LookupEnv("https_proxy"); ok {
```
Comment (isubasinghe) [inline workflow/artifacts/oss/oss.go:88]: Can you add a comment that this is needed for the OSS driver and that the S3 and GCS drivers respect this environment variable?

**Review by isubasinghe — CHANGES_REQUESTED** [review]

File: workflow/artifacts/oss/oss.go (hunk 1 above)
Comment (jingkkkkai) [reply workflow/artifacts/oss/oss.go:90]: I have tested other drivers and added doc for oss driver.
please refer #12313 and [oss cli doc](https://help.aliyun.com/zh/cli/use-an-http-proxy-server#section-5yf-ejl-jwf)
thanks

**Comment by carolkao** [comment]
Hi @terrytangyuan and @isubasinghe,
I'd like to expedite the progress of this PR. It's crucial for supporting proxy settings in OSS client for us. Currently, we rely on Argo v3.4.8 with our self-fix build in our production environment. However, this setup is blocking any of our Argo Workflow upgrade plans, including trying out the new build which addresses the potential issue detailed in https://github.com/argoproj/argo-workflows/issues/11948.
I hope you can help move the review forward. Thank you so much.

**Comment by shuangkun** [comment]
Could you solve the test error？

**Comment by carolkao** [comment]
Hi @jingkkkkai , please help to resolve the error for this PR.

**Comment by isubasinghe** [comment]
@jingkkkkai DCO isn't passing and the test suite isn't passing, maybe try commit an empty commit.

**Comment by jingkkkkai** [comment]
Hi @shuangkun , @isubasinghe 
the test suite passed after adding an empty commit, thanks for your help
then could you please help review the pr again?

additionally, is there a way to apply this change to branch v3.4?

**Comment by carolkao** [comment]
Hi @isubasinghe,
All checks have passed now, can you help to move this PR forward?

**Review by isubasinghe — APPROVED** [review]
LGTM

**Comment by carolkao** [comment]
Hi @terrytangyuan ,

Do you think Argo's development team could help cherry-pick this fix into the 3.4 branch? We're struggling with the build environment, and I believe it would be much easier for Argo's team to handle this.

Thanks,
Carol

**Review by terrytangyuan — APPROVED** [review]

**Comment by terrytangyuan** [comment]
@carolkao See https://github.com/argoproj/argo-workflows/issues/11851#issuecomment-2339509093

**Comment by carolkao** [comment]
> @carolkao See [#11851 (comment)](https://github.com/argoproj/argo-workflows/issues/11851#issuecomment-2339509093)

Cool, thank you @terrytangyuan . Looking forward to the next 3.4 release.

**Pull request merged**